### PR TITLE
Append the caller data to each SQL query

### PR DESCRIPTION
**Pull request by sewi-cpan**
This PR appends an SQL comment to each query containing the Python function/file/line
calling Django to run that command.

We experienced a busy database server blocked by one very expensive query. Another
situation caused too many connections to the database server. Identifying the troublemaker
within the source code wasn't easy.

This PR adds the caller of each query to `SELECT query FROM pg_stat_activity` as a SQL comment:
```
INSERT INTO "tests_47ee19d1" ("id", "title")
VALUES (1, 'Test')
RETURNING "tests_47ee19d1"."id"
/* 998020 test_append_caller_to_sql_crud .../django-postgres-extra/tests/test_append_caller_to_sql.py 55 */

SELECT "tests_47ee19d1"."id", "tests_47ee19d1"."title"
FROM "tests_47ee19d1"
WHERE "tests_47ee19d1"."id" = 1
LIMIT 1
/* 998020 test_append_caller_to_sql_crud .../django-postgres-extra/tests/test_append_caller_to_sql.py 69 */

UPDATE "tests_47ee19d1"
SET "title" = 'success'
WHERE "tests_47ee19d1"."id" = 1
/* 998020 test_append_caller_to_sql_crud .../django-postgres-extra/tests/test_append_caller_to_sql.py 64 */

DELETE FROM "tests_47ee19d1"
WHERE "tests_47ee19d1"."id" IN (1)
/* 998020 test_append_caller_to_sql_crud .../django-postgres-extra/tests/test_append_caller_to_sql.py 74 */
```
The comment contains:

1. The system process id (PID). `pg_stat_activity` has a column `client_addr`. The server IP and the process id let you track down the originating process, even if PgBouncer is used inbetween.

2. The function name of the last function or method calling Django.

3. The file name and path of the source file where the call to Django happened.

4. The line number within the source file.

Djangos internal commands show the process ID and name for easy identification of migrations and other maintenance operations.

[#181799346]

### CODE REVIEW

File: psqlextra/compiler.py [hunk 1]
```diff
@@ -6,13 +10,59 @@
...
+def append_caller_to_sql(sql):
+    try:
+        # Search for the first non-Django caller
+        stack = inspect.stack()
+        i = 0
+        for stack_frame in stack[1:]:
+            i += 1
+            frame_filename = stack_frame[1]
+            frame_line = stack_frame[2]
+            frame_function = stack_frame[3]
```
Comment (Photonios) [inline psqlextra/compiler.py:35]: **suggestion (non-blocking)**:

```suggestion
        for i, frame in enumerate(stack[1:]):
             frame_filename, frame_line, frame_function, _, _, _ = frame
```

File: psqlextra/compiler.py [hunk 2]
```diff
@@ -72,19 +123,32 @@ def _does_dict_contain_expression(data: dict) -> bool:
...
         self.qn = self.connection.ops.quote_name
 
     def as_sql(self, return_id=False):
         """Builds the SQL INSERT statement."""
-
         queries = [
             self._rewrite_insert(sql, params, return_id)
             for sql, params in super().as_sql()
         ]
+        print(f"InsertConflictCompiler {queries}")
```
Comment (Photonios) [inline psqlextra/compiler.py:151]: **issue (blocking)**: Can be removed.

File: psqlextra/compiler.py [hunk 3]
```diff
@@ -132,9 +196,12 @@ def _rewrite_insert(self, sql, params, return_id=False):
...
 
-        return self._rewrite_insert_on_conflict(
-            sql, params, self.query.conflict_action.value, returning
-        )
+        if hasattr(self.query, "conflict_action"):
+            (sql, params) = self._rewrite_insert_on_conflict(
+                sql, params, self.query.conflict_action.value, returning
+            )
+
+        return (append_caller_to_sql(sql), params)
```
Comment (Photonios) [inline psqlextra/compiler.py:204]: **suggestion (non-blocking)**:

```suggestion
        return append_caller_to_sql(sql), params
```

File: psqlextra/compiler.py [hunk 4]
```diff
@@ -132,9 +196,12 @@ def _rewrite_insert(self, sql, params, return_id=False):
             self.qn(self.query.model._meta.pk.attname) if return_id else "*"
         )
 
-        return self._rewrite_insert_on_conflict(
-            sql, params, self.query.conflict_action.value, returning
-        )
+        if hasattr(self.query, "conflict_action"):
```
Comment (Photonios) [inline psqlextra/compiler.py:199]: **question:** Why was this extra check added?

File: psqlextra/compiler.py [hunk 5]
```diff
@@ -72,19 +123,32 @@ def _does_dict_contain_expression(data: dict) -> bool:
...
+        """Builds the SQL INSERT statement."""
+        queries = [
+            (append_caller_to_sql(sql), params)
+            for sql, params in super().as_sql()
+        ]
+
+        return queries
+
 
+class PostgresInsertOnConflictCompiler(SQLInsertCompiler):
```
Comment (Photonios) [inline psqlextra/compiler.py:136]: **question**: Why was this renamed and splitted?

File: psqlextra/compiler.py [hunk 6]
```diff
@@ -6,13 +10,59 @@
...
+    SQLInsertCompiler,
+    SQLUpdateCompiler,
+)
 from django.db.utils import ProgrammingError
 
 from .expressions import HStoreValue
 from .types import ConflictAction
 
 
+def append_caller_to_sql(sql):
```
Comment (Photonios) [inline psqlextra/compiler.py:26]: **question**: Can we put this behind a Django setting flag, disabled by default? Not all users might want this feature enabled by default.

File: psqlextra/compiler.py (hunk 4 above)
Comment (sewi-cpan) [reply psqlextra/compiler.py:199]: You're right. Has been a leftover from the first attempt to use the conflict handler for all "INSERT" statements and is removed now.

File: psqlextra/compiler.py (hunk 5 above)
Comment (sewi-cpan) [reply psqlextra/compiler.py:136]: The original `PostgresInsertCompiler` only applied to `INSERT ... ON CONFLICT` queries and is incompatible with `INSERT` queries without `ON CONFLICT` condition. This has been invisible here and has been covered by `PostgresOperations.compiler` by preferring the Django-detected compiler. The change just made that already existing split visible and added a simple wrapper for plain `INSERT`s.

**Review by Photonios — APPROVED** [review]
